keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,645 @@
+		  toVersion: v1
+		  rules:
+		    - field: spec.xyz.podIPs
+			  rule: "has(spec.xyz.podIP) ? [spec.xyz.podIP] : []"
+		- fromVersion: v1
+		  toVersion: v1beta1
+		  rules:
+		    - field: xyz.podIP
+		      scope: spec
+			  rule: "size(xyz.podIPs) > 0 ? xyz.podIPs[0] : null"
+```
+
+Expressions that return object types will overwrite all fields that they return and
+leave all other fields to be auto-converted.
+
+The 'scope' is optional and defaults to the object root.
+
+The 'field' is optional and defaults to the scope? TODO: is it better to make it required for conversion?
+
+If the expression evaluates to null, the field is left unset.
+
+TODO: demonstrate writing fields into annotations for round trip
+
+TODO: demonstrate string <-> structured (label selector example & ServicePort example)
+
+#### Other validation support
+
+  Add a `type` field specified with `cel` within `x-kubernetes-validator`.
+  With `type` field added, the validator could potentially add other ways to validate in
+  the future, or to allow 3rd party validators to have a way to inline their
+  validation rules in CRDs.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Type Checking
+
+CEL type checking requires "declarations" be registered for any types that are to
+be type checked.  In our case, the type information of interest is the CRD's
+structural schemas. So we need to translate structural schemas to declarations.
+
+"declarations" are registered via the go-genproto "checked" package
+(https://github.com/googleapis/go-genproto/blob/master/googleapis/api/expr/v1alpha1/checked.pb.go).
+
+(https://github.com/googleapis/googleapis/blob/master/google/api/expr/v1alpha1/checked.proto).
+
+There are a couple alternative ways to do this. Ideally, we could be able to both dereference into
+objects and construct objects in a typesafe way.  In order to construct objects in a typesafe way we
+need to be able to represent the structural schema types in CEL, e.g. "v1beta1.Foo{fieldname:
+value}", this is complicated by the way CEL relies on protobuf types.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: CustomResourceValidationExpressions
+  - Components depending on the feature gate: kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No, default behavior is the same.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, disabling the feature will result in validation expressions being ignored.
```

Comment:
Can I get a commitment to a unit test that ensures that if the featuregate is off, but the field is present, an update which does not add more will not fail?

Reply:
Absolutely. KEP is updated to match.